lib/spack/spack/config.py

```diff
@@ -1306,6 +1311,79 @@ def _config_from(scopes_or_paths):
     return configuration
 
 
+def raw_github_gitlab_url(url):
+    """Transform a github URL to the raw form to avoid undesirable html.
+
+    Args:
+        url: url to be converted to raw form
+
+    Returns: (str) raw github/gitlab url or the original url
+    """
+    if 'github' in url:
+        # Github provides access to raw file contents (i.e., no html tags)
+        # at the following URL.
+        raw_url = url.replace('github.com/', 'raw.githubusercontent.com/')
+
+        # Generally develop URLs have a 'blob' directory on github.com so
+        # drop it if present.
+        raw_url = raw_url.replace('/blob/', '/')
+
+        # URLs can sometimes redirect(?) to a 'blame' location that can be
+        # dropped if present to get the raw contents.
+        return raw_url.replace('/blame/', '/')
+
+    if 'gitlab' in url:
+        # fortunately appears gitlab just uses a different directory
+        return url.replace('/blob/', '/raw/')
+
+    return url
+
+
+def fetch_remote_configs(url, dest_dir):
+    """Retrieve configuration file(s) at the specified URL.
+
+    Arguments:
+        url (str): URL for a configuration (yaml) file or a directory containing
+            yaml file(s)
+        dest_dir (str): destination directory
+
+    Returns: (str) path to the corresponding file if URL is or contains a
+       single file or the root (dest_dir) directory if multiple configuration
+       files are present
+    """
+    def fetch_content(url):
+        root, ext = os.path.splitext(url)
+        if ext == '.yaml':
+            raw = raw_github_gitlab_url(url)
+            tty.debug("Reading config from url {0}".format(raw))
+            return spack.url.fetch_url_text(raw, dest_dir=dest_dir)
+
+        return None
+
+    if not url:
+        raise ConfigFileError("Cannot retrieve configuration without a URL")
+
+    # Return the local path if the fetched content corresponds to a single file
+    path = fetch_content(url)
+    if path:
+        return path
+
+    # Check if the url is a "directory" containing one or more configuration
+    # files and return the corresponding local path of the fetched file(s).
+    pages = web_util.spider(url, 0)
```

Comment:
This should assign the return value to a tuple like

```
pages, links = web_util.spider(url, 0)
```

or

```
_, links = web_util.spider(url, 0)
```

and regarding https://github.com/spack/spack/pull/29026#discussion_r881147374

> Please clarify there should probably be some validation to ensure this. spack.util.web.spider will return the HTML page and associated links -- including the modules.yaml file -- for https://github.com/spack/spack/tree/develop/etc/spack/defaults.

My understanding is that this should create a directory consisting of YAML files, as it is it seems like this can download non-YAML files to a directory. The fact that the directory contains non-YAML files isn't strictly problematic since I think the config logic will ignore them, but IMO it's worth avoiding a download for files we don't intend to use.

Reply:
Thanks.  Let me revisit the download aspect as there are a LOT of irrelevant links getting returned from a GitHub directory with one or more `.yaml` files.

Reply:
FYI. `web_util.spider(url, 0)` returns the `html` for the `url` as a dictionary entry in `pages` and *all* detected URLs within it in `links`.  Extracting the text from the `html` output is not desirable.  The URLs in `links` have to be filtered since they may contain many irrelevant URLs.  For example, https://github.com/spack/spack/blob/develop/etc/spack/defaults/bootstrap.yaml has 70 links, some of which are:

```
https://github.com/spack/spack/pulse
https://github.com/login?return_to=%2Fspack%2Fspack
https://github.com/pricing#compare-features
https://github.com/spack/spack/tree/develop/etc/spack
https://github.com/spack/spack/blame/develop/etc/spack/defaults/bootstrap.yaml
https://github.com/spack
https://github.com/features/codespaces
https://github.com/features/code-review
https://github.com/spack/spack/blame/c36f15e29eff5d1f9cb6f7932a0edd7c420bf9e9/etc/spack/defaults/bootstrap.yaml
[..snip..]
https://github.com/spack/spack/blob/c36f15e29eff5d1f9cb6f7932a0edd7c420bf9e9/etc/spack/defaults/bootstrap.yaml
https://github.com/spack/spack/blob/develop/etc/spack/defaults/#start-of-content
[..snip..]
https://github.com/team
https://github.com/mobile
https://github.com/spack/spack/raw/develop/etc/spack/defaults/bootstrap.yaml
https://docs.github.com/en/github/site-policy/github-privacy-statement
https://github.com/enterprise
```